### ISSUE TITLE
ABP CLI

### ISSUE THREAD

**Issue by hikalkan** [abpframework/abp#544]
### Overview

Initially, this tool will have some functionalities:

1. Create a new solution using ABP.
2. Add a module to an existing solution.
3. Upgrade ABP modules/packages easily.

### Details

A draft example command to create a new solution:
````
abp new Acme.PhoneBook -orm EF
abp new Acme.PhoneBook -orm EF -ui Angular
````

A draft example command to add a new application module to an existing solution (.sln):
````
abp add-module Blogging
abp add-module Blogging -orm MongoDB
````
- It should be executed at the folder of the .sln file.
- This example adds Blogging module to the solution.
- Blogging module (somehow) define how to add itself to an existing solution respecting to the layering.
- It also adds [DependsOn(...)] attributes to related module definition classes.
- It may make additional changes in the codebase (for example, adds itself to EF Core migration dbcontext, run add-migration command...)

>Note that a module (like Blogging) may contain multiple nuget packages.

A draft example command to add a new package to an existing project (.csproj):
````
abp add-module Volo.Abp.RabbitMQ
````
- It should be executed at the folder of a .csproj file.
- This example adds Volo.Abp.RabbitMQ package to the project as a nuget package.
- It also adds [DependsOn(...)] attribute to the module definition class in the project.
- It may make additional changes in the codebase (for example, adds itself to EF Core migration dbcontext, run add-migration command... if defined)

### Some Resources

* .NET Core Global Tools overview: https://docs.microsoft.com/dotnet/core/tools/global-tools
* An example implementation for [ASP.NET Boilerplate](https://aspnetboilerplate.com/): https://github.com/personball/Abp-CLI

**Comment by hey-hoho** [comment]
it's really wonderful.

**Comment by tairan** [comment]
Can install the apb template into dotnet new command via `dotnet new -i Abp.WebMVC`

Just like these third part template https://github.com/dotnet/templating/wiki/Available-templates-for-dotnet-new

**Comment by maliming** [comment]
https://github.com/natemcmaster/CommandLineUtils
Here is a tool for command line parsing. Looks very good.

**Comment by personball** [comment]
@hikalkan  Consider https://github.com/personball/Abp-CLI

**Comment by hikalkan** [comment]
Thanks @personball for pointing this

**Comment by hikalkan** [comment]
This is done and [already documented](https://github.com/abpframework/abp/blob/master/docs/en/CLI.md). Will be available with new features with the v0.18.
